src/UrlGenerator.php

```diff
@@ -0,0 +1,224 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Router\FastRoute;
+
+use Nyholm\Psr7\Uri;
+use Yiisoft\Router\RouteCollectionInterface;
+use Yiisoft\Router\RouteNotFoundException;
+use Yiisoft\Router\UrlMatcherInterface;
+use Yiisoft\Router\UrlGeneratorInterface;
+use FastRoute\RouteParser;
+
+use function array_key_exists;
+use function array_keys;
+use function implode;
+use function is_string;
+use function preg_match;
+
+final class UrlGenerator implements UrlGeneratorInterface
+{
+    private string $uriPrefix = '';
+    private RouteCollectionInterface $routeCollection;
+    private UrlMatcherInterface $matcher;
+    private RouteParser $routeParser;
+
+    public function __construct(
+        UrlMatcherInterface $matcher,
+        RouteParser $parser = null
+    ) {
+        $this->matcher = $matcher;
+        $this->routeCollection = $matcher->getRouteCollection();
```

Comment:
`$matcher` is type-hinted as `UrlMatcherInterface` but `getRouteCollection()` is not part of the interface. Either we should type-hint against `UrlMatcher` or add the method to the interface (which isn't a good idea since it's implementation detail).

Reply:
It will be added to the interface.